### PR TITLE
Change @experimental spec

### PR DESCRIPTION
**Pull request by nicolasstucki**
The main differences with the current version is that @experimental is not experimental and that we are able to define experimental APIs using any version of the compiler (nightly, snapshot and stable release).

See ` docs/docs/reference/experimental/experimental-defs.md`.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 1]
```diff
@@ -948,27 +948,23 @@ object RefChecks {
...
-    if !sym.owner.isExperimental && !pos.span.isSynthetic then // avoid double errors
-      checker.traverse(sym.info)
+    if !sym.isInExperimentalScope then
+      new Checker().traverse(sym.info)
 
   private def checkExperimentalAnnots(sym: Symbol)(using Context): Unit =
-    for annot <- sym.annotations if annot.symbol.isExperimental && annot.tree.span.exists do
-      Feature.checkExperimentalDef(annot.symbol, annot.tree)
+    if !sym.isExperimental then
+      for annot <- sym.annotations if annot.symbol.isExperimental && annot.tree.span.exists do
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:966]: Would be good to add a comment explaining why we check for tree.span.exists.

File: docs/docs/reference/other-new-features/experimental-defs.md [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+  p3: X,
+  p4: Y,
+  p5: Z,
+  p6: Any = x
+) {}
+```
+
+</details>
+
+(3.) The extension clause of an experimental `class`, `trait`, `object` are experimental scopes.
```
Comment (smarter) [inline docs/docs/reference/other-new-features/experimental-defs.md:180]:
```suggestion
(3.) The `extends` clause of an experimental `class`, `trait` or `object` is an experimental scope.
```

File: docs/docs/reference/other-new-features/experimental-defs.md [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+@experimental class B1 extends A1(1)
+class B2 extends A1(1) // error: class A1 is marked @experimental and therefore marked @experimental and therefore ...
+
+@experimental class C1 extends A2(x)
+class C2 extends A2(x) // error def x is marked @experimental and therefore
+```
+
+</details>
+
+(4.) Members of an experimental `class`, `trait` or `object` are in experimental scopes.
```
Comment (smarter) [inline docs/docs/reference/other-new-features/experimental-defs.md:203]: Might be simpler to state that as:
```suggestion
(4.) The body of an experimental `class`, `trait` or `object` is an experimental scope.
```

File: library/src-non-bootstrapped/scala/annotation/internal/experimentalTest.scala [hunk 4]
```diff
@@ -0,0 +1,10 @@
+// KEEP IN NON-BOOTSTRAPPED SOURCES
+package scala.annotation
+package internal
+
+/** This is a dummy definition that tests that the stdlib can be compiled with an experimental definition.
+ *  This might be redundant, we keep it definition in case there are no other @experimental definitions in the library.
+ *  As this definition is in `src-non-bootstrapped`, it will not be published.
+ *  It may accidentally be visible while compiling the non-bootstrapped library.
+ */
+@experimental def testExperimental = 4
```
Comment (smarter) [inline library/src-non-bootstrapped/scala/annotation/internal/experimentalTest.scala:10]: Not sure how important it is to test this given that `@experimental` is no longer experimental itself now?

**Review by smarter — APPROVED** [review]

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/typer/RefChecks.scala]: THat condition is not required anymore. I removed it.

File: library/src-non-bootstrapped/scala/annotation/internal/experimentalTest.scala (hunk 4 above)
Comment (nicolasstucki) [reply library/src-non-bootstrapped/scala/annotation/internal/experimentalTest.scala:10]: This is not really there to check if `@experimental` is experimental, it is to check that we can define experimental definitions in the standard library. This is to make sure we catch a regression as early as possible as this kind of failure could block an update if the reference compiler.

Note that for `3.1.0` we will not have any other experimental definitions. They will all get stabilized.

**Review by odersky — APPROVED** [review]

**Pull request merged**